vignettes/qenv.Rmd

```diff
@@ -0,0 +1,124 @@
+
+```{r}
+library(magrittr)
+
+q2 <- eval_code(my_qenv, "y <- x * 2") %>% eval_code("z <- y * 2")
+
+# my_qenv still contains only x
+print(my_qenv)
+
+# q2 contains x, y and z
+print(q2)
+```
+
+Objects can be extracted from a `qenv` using `[[` for example in order for them to be displayed in a shiny app. The code used to generate the `qenv` is obtained using the `get_code()` function.
+
+```{r}
+
+print(q2[["y"]])
+
+cat(paste(get_code(q2), collapse = "\n"))
+```
+### Joining `qenv` objects
+
+Given a pair of `qenv` objects it might be possible to "join" them together creating a new `qenv` object containing the union of both environments together with the code needed to reproduce it:
+
+```{r}
+
+common_q <- eval_code(new_qenv(), quote(x <- 1))
+
+x_q <- eval_code(common_q, quote(y <- 5))
+y_q <- eval_code(common_q, quote(z <- 5))
+
+join_q <- join(x_q, y_q)
+
+print(join_q)
+```
+
+Dependent on the contents of the environments and order of code it may not be possible to join `qenv` objects. See the function documentation for more details.
+
+## Using `qenv` inside `shiny` applications
+
+The functions above can be combined into a `shiny` application generating reproducible outputs. In the example below the `verbatim_popup` from `teal.widgets` is used to hide the code used to generate output in a modal dialog. 
+When using a `qenv` to evaluate code, if an error occurs then an object of type `qenv.error` is created. This object can be used in all the same places a `qenv` object is used and means you do not need to change your code to handle these errors - select `error_option` in the example below to see `qenv` error handling in action. 
+
+```{r}
+library(shiny)
+library(teal.widgets)
+
+# create an initial qenv with the data in
+data_q <- new_qenv() %>% eval_code("iris_data <- iris")
+
+ui <- fluidPage(
+  radioButtons(
+    "option", "Choose a column to plot:",
+    c("Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width", "error_option")
+  ),
+  # button for the Code to be shown when clicked on
+  verbatim_popup_ui("rcode", button_label = "Show Code"),
+  plotOutput("plot")
+)
+
+server <- function(input, output, session) {
+
+  # create a qenv containing the reproducible output
+  output_q <- reactive({
+    req(input$option)
+    eval_code(
+      data_q,
+      bquote(p <- hist(iris_data[, .(input$option)]))
+    )
+  })
+
+  # display output
+  output$plot <- renderPlot(output_q()[["p"]])
+
+  # code is displayed when "Show Code" button is clicked on
+  verbatim_popup_srv("rcode", reactive(paste(get_code(output_q()), collapse = "\n")), title = "Code")
+}
+
+if (interactive()) {
+  shinyApp(ui, server)
+}
+```
+
+
+## `qenv` and `teal` applications
+
+The `qenv` object can easily be incorporated into `teal` modules. See the `teal` vignette [Creating Custom Modules](https://insightsengineering.github.io/teal/main/articles/creating-custom-modules.html) for further details.
```

Comment:
This vignette has been updated on the teal_refactor@main branch to include qenv see [here](https://github.com/insightsengineering/teal/blob/teal_refactor%40main/vignettes/creating-custom-modules.Rmd#L138)